[Yu-chuan Liao] iP

isanidiot opened the pull request:
### UNC

> "I hate Duke with an infernal passion undying. I hate every leaf of every tree on that sickening campus." -Ian Williams, UNC alumnus
UNC is a simple chat bot that tracks your tasks. It's,

- Text-based
- Easy to learn
- ~~Fast~~ *SUPER FAST* to use!

All you need to do is,

1. Download it from [here](https://github.com/isanidiot/ip/releases/tag/v0.1)
2. Copy the jar file into an empty folder
3. Open a command window in that folder
4. Run the command java -jar unc.jar
5. Add your tasks
6. Let it manage your tasks for you 🉐 

And it's **FREE!**

Features:

- [X] Managing tasks
- [X] Managing Deadlines
- [X] Managing Events
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public static void main(String[] args) {
        new Unc("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

isanidiot committed:
Level 1 greet, echo, exit

isanidiot committed:
Level 2 add, list

isanidiot committed:
Level 3 mark; Extension class

isanidiot committed:
Level 4 todo, event, deadline; Extension inheritance

isanidiot committed:
Extension text ui testing

isanidiot committed:
Level 5 handle errors; Extension exceptions

isanidiot committed:
Level 6 delete; Extension collections

isanidiot committed:
Extension enums

isanidiot committed:
Level 7 save

isanidiot committed:
Level 8 dates and times

isanidiot committed:
Merge branch 'branch-Level-7'

isanidiot committed:
Merge branch 'branch-Level-8'

isanidiot committed:
handle merge conflicts

isanidiot committed:
Extension more oop

isanidiot committed:
Extension packages

isanidiot committed:
Merge branch 'add-gradle-support'

isanidiot committed:
Extension gradle

isanidiot committed:
Extension junit

isanidiot committed:
Extension jar

isanidiot committed:
Extension javadoc

isanidiot committed:
Extension coding standard

isanidiot committed:
Level 9 find

isanidiot committed:
Merge branch 'branch-A-JavaDoc'

isanidiot committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/unc/Ui.java

isanidiot committed:
Merge branch 'branch-Level-9'

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+package unc;
```
Great job for week 3! However, there is room of improvement to improve code quality. Keep up the good work 😄

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+        DEADLINE,
+        EVENT
+
+    }
```
Could this be separated into another Enum file? It can reduce duplicated codes across the different classes 🤔

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+                            System.out.println("Error creating saved list.");
+                        }
+                        break;
+                    default:
```
Should there be an indentation before every "case"?

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+                saveFile.createNewFile();
+            } catch (IOException ioException) {
+                System.out.println(ioException);
+            } catch (SecurityException securityException) {
```
Nice catch for Security Exceptions here! 👍

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+
+        String input;
+        Scanner scanner = new Scanner(System.in);
+        while(!Objects.equals(input = scanner.nextLine(), "bye")) {
```
Feels a bit difficult to read and understand for me.
Could this be separated into multiple lines for better readability?

cadencjk commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+        taskList.add(new unc.task.Task(input));
+    }
+
+    */
```
Should this draft code be deleted?

cadencjk commented on src/main/java/unc/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Abstract class of executable commands.
+ */
+abstract public class Command {
```
Nice use of abstract classes here! 👍

cadencjk commented on src/main/java/unc/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,48 @@
+package unc.command;
+
+import unc.*;
```
Will it be better if individual packages are imported rather than using a wildcard?

cadencjk reviewed (commented):
Comments are detailed and pretty well written. 
More code could be abstracted into a parent class to reduce redundancy.
LGTM! 😄

beetee17 commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.*;
```
Could consider breaking up this import statement into multiple explicit statements.

beetee17 commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+        taskList = new ArrayList<>();
+        saveFile = new File("./data/save.txt");
+        try {
+            Scanner scanner1 = new Scanner(saveFile);
```
Perhaps you could consider choosing a more specific name for the scanner variables since you have two in the same scope, rather than `scanner` and `scanner1`?

beetee17 commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+    private static List<Task> taskList;
+    private static File saveFile;
+
+    private enum validInput {
```
Should the enum be in Pascal case instead (i.e. `ValidInput`)?

beetee17 commented on src/main/java/unc/Duke.java:
```diff
@@ -0,0 +1,198 @@
+
+    */
+
+    private static void addToDo(String input) throws UncException {
```
This is just nitpicking, but you could consider renaming the method to `addTodo`, especially since the class name is not capitalised in this way as well

beetee17 commented on src/main/java/unc/Storage.java:
```diff
@@ -0,0 +1,101 @@
+package unc;
+
+import unc.task.Deadline;
+import unc.task.Event;
+import unc.task.Task;
+import unc.task.Todo;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
Nice job separating all the import statements into explicit ones here!

beetee17 commented on src/main/java/unc/UncException.java:
```diff
@@ -0,0 +1,12 @@
+package unc;
+
+public class UncException extends Exception{
```
You might've forgotten a space here! 

P.S I noticed this in other places as well e.g `EventCommand.java`. You may want to consider using checkstyle to help you catch these errors automatically! Check out [the guide here](https://se-education.org/guides/tutorials/checkstyle.html)

beetee17 reviewed (commented):
Just a few minor lapses that can be easily fixed with checkstyle, keep it up!

isanidiot committed:
Extension checkstyle

isanidiot committed:
Merge branch 'branch-A-CheckStyle'

isanidiot committed:
Level 10 gui

isanidiot committed:
Merge branch 'branch-Level-10'

isanidiot committed:
Extension assertions

There are some values in our program that should never be null whether the user input is valid or not.

Using assertions helps verify the assumptions that they are not null.

Let's add some assertions to make sure our tasks and storage are initialised correctly.

isanidiot committed:
Extension code quality

There were several lines of dead code, as well as a header comment that didn't follow the suggested format.

Improving the code quality can make it easier to read and debug the code.

Let's remove the dead code and edit the JavaDoc.

isanidiot committed:
Merge pull request #2 from isanidiot/branch-A-Assertions

Extension assertions

isanidiot committed:
Merge branch 'master' of https://github.com/isanidiot/ip

isanidiot committed:
Merge branch 'master' into branch-A-CodeQuality

isanidiot committed:
Merge pull request #3 from isanidiot/branch-A-CodeQuality

Branch a code quality

isanidiot committed:
Extension sort

isanidiot committed:
Merge branch 'branch-C-Sort'

isanidiot committed:
bug fix: storage; javadoc: findcommand, sortcommand

isanidiot committed:
Update README.md

isanidiot committed:
Update README.md

isanidiot committed:
Update README.md

isanidiot committed:
Update README.md

isanidiot committed:
handle exceptions

isanidiot committed:
Merge remote-tracking branch 'origin/master'

isanidiot committed:
Ui.png
